### PR TITLE
It is time Simba!

### PR DESCRIPTION
**Pull request by jboucher154**
Les move our current product to the main branch!

### CODE REVIEW

**Review by azarSarikhani — APPROVED** [review]
Great small step for mankind, magnificent giant step for webServants

File: README.md [hunk 1]
```diff
@@ -1 +1,154 @@
-# WebServ
+# WebServ: JAS Server
+
+A simple HTTP server written in C++
+
+This project introduced us to the wonderful world of HTTP, RFC, server configurations, CGI, and TCP socket mangement. 
+This simple server provides GET, HEAD, POST, and DELETE methods and provides common gateway interface (CGI) for scripted/  interpred languages such as python and bash. It was developed to specifically work with a chrome client and run on Mac OS 10.14.6.
```
Comment (sakarisalmi) [inline README.md:6]: line 6: interpreted, Capital C on chrome.

File: README.md [hunk 2]
```diff
@@ -1 +1,154 @@
...
+
+[Azär Sarikhani](https://github.com/azarSarikhani)
+
+[Sakari Salmi](https://github.com/sakarisalmi)
+
+## Project Coding Standards
+
+Coding standards set for the project can be found [here](project_standards/standards.md)
+
+The project began with guidelines to use C++98 standard. The limit was lifted towoards the end of the project so no major refactoring was done, yet a few methods and objects from standards beyond 98 can be found.
```
Comment (sakarisalmi) [inline README.md:22]: "to use with c++98 standard", towards, so -> but, beyond c++98

File: README.md [hunk 3]
```diff
@@ -1 +1,154 @@
...
+- Mac OS 10.14.6
+- Bash
+- Python
+
+## Features
+
+### General
+
+- Serves fully static websites
+- Follows HTTP 1.1, though is not fully compliant
```
Comment (sakarisalmi) [inline README.md:37]: "though not fully compliant"

File: README.md [hunk 4]
```diff
@@ -1 +1,154 @@
...
+- Each block must be enclosed in curley braces `{}`.
+- Each  line in a `main` or `location` block must end with a `;`.
+- Comments can be made in the config file using a leading `#`.
+
+#### Basic Requirements
+
+The config file requires a `server` block to contain at least a `main` block and a `location` block for the root (`/`). 
+
+The `main` block must have the following directives designated:
+- listen: value is the port the server will listen on
```
Comment (sakarisalmi) [inline README.md:53]: "value of the port the server will listen to"

File: README.md [hunk 5]
```diff
@@ -1 +1,154 @@
...
+- server_name: name of the server (Note: If ports are shared the server_name must be unique.)
```
Comment (sakarisalmi) [inline README.md:54]: "(Note: If ports are shared, the server_name must be unique.)"

File: README.md [hunk 6]
```diff
@@ -1 +1,154 @@
...
+- host: the IP of the server
+- root: the path to the root of the served content (will be used for locations without given roots)
+- client_max_body_size: largest size of body in request that the server will accept. Limited to size of double.
+
+The `main` block may have error pages designated with this format
+- error_page_404 followed/by/path/to/error_page.html;
+
+If nothing is designated for a `location` block the following defaults will be set:
```
Comment (sakarisalmi) [inline README.md:62]: If nothing is designated for a `location` block, the following defaults will be set:

File: README.md [hunk 7]
```diff
@@ -1 +1,154 @@
...
+- root: will use the root from the `main` block if none given
+
+
+The following are optional `location` directives that will not have a default set:
+- return: set to the name of another location for a redirection to be sent to client
+- alias: set to name of location for server to refence for resources, other directive set here will override directive in the other location.
+
+At least one directive needs to be set for a `location` to be valid.
+
+Special location `/cgi-bin` can be designated to configure CGI script. It must have this name and has some unique directives to be set.
```
Comment (sakarisalmi) [inline README.md:75]: separate this into two different sentences.

**Review by sakarisalmi — COMMENTED** [review]
Thanks for your patience.
These fixes aren't vital, but it was fun to spell check

File: README.md (hunk 1 above)
Comment (jboucher154) [reply README.md:6]: done

File: README.md (hunk 2 above)
Comment (jboucher154) [reply README.md:22]: I rewrote it with grammarly.

File: README.md (hunk 3 above)
Comment (jboucher154) [reply README.md:37]: done

File: README.md (hunk 4 above)
Comment (jboucher154) [reply README.md:53]: done

File: README.md (hunk 5 above)
Comment (jboucher154) [reply README.md:54]: done

File: README.md (hunk 6 above)
Comment (jboucher154) [reply README.md:62]: done

File: README.md [hunk 8]
```diff
@@ -1 +1,154 @@
...
+- root: the path to the root of the served content (will be used for locations without given roots)
+- client_max_body_size: largest size of body in request that the server will accept. Limited to size of double.
+
+The `main` block may have error pages designated with this format
+- error_page_404 followed/by/path/to/error_page.html;
+
+If nothing is designated for a `location` block the following defaults will be set:
+- index: will be set to index.html for the default file
+- allow_methods: default will be no methods allowed
+- autoindex: default will be off, no directory listing will be shown the client will be directed to the default file set in the index directive
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:65]: - autoindex: default will be `off`, in which case no directory listing will be shown and the client will be directed to the default file set in the index directive

File: README.md [hunk 9]
```diff
@@ -1 +1,154 @@
...
+
+The following are optional `location` directives that will not have a default set:
+- return: set to the name of another location for a redirection to be sent to client
+- alias: set to name of location for server to refence for resources, other directive set here will override directive in the other location.
+
+At least one directive needs to be set for a `location` to be valid.
+
+Special location `/cgi-bin` can be designated to configure CGI script. It must have this name and has some unique directives to be set.
+- script: list of scripts existing in the /cgi-bin location
+- cgi_path: path to the interpreter/ executor for the scripts in same order as the extensions (below) e.g, /usr/local/bin/python3 /bin/bash
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:77]: done

File: README.md [hunk 10]
```diff
@@ -1 +1,154 @@
...
+
+### HTTP Status Codes
+
+HTTP Status Codes are designated my the RFC and used to communicate specific messages of Information (1XX), Success (2XX), Redirection (3XX), Client Errors (4XX), and Server Error (5XX).
+
+### HTTP Request/ Response
+
+General notes on request and response handling by the JAS Server:
+- The server honours reactive negotiation request/ response pattern
+- Chunked requests will be unchunked. Once all the chunks are received the request will be processed and a response made
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:89]: done

File: README.md [hunk 11]
```diff
@@ -1 +1,154 @@
...
+- Currently Responses are not sent as chunked.
+- If not all parts of a request have not been recieved in the REQUEST_TIMEOUT_SEC (default set to 10) a timeout response will be sent to client and the connection closed.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:91]: done

File: README.md [hunk 12]
```diff
@@ -1 +1,154 @@
...
+
+### HTTP Methods
+
+The JAS Server provides GET, HEAD, POST, DELETE methods. Any other method will return a `501 NOT_IMPLEMENTED` response.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:95]: I refuse. Oxford commas are always needed.

File: README.md [hunk 13]
```diff
@@ -1 +1,154 @@
...
+
+#### GET
+
+GET method is safe and idempotent. If GET method is allowed by the location for which the request URI is requesting the resource from, the resource will be returned in the body of the response. This of course depends on the resource being found and accessable.
+
+Due to project limitations, the size of media that is able to be returned may be limited depending on the system that is running it, especially if the resource is larger than 5MB. In the case of an incomplete response sent an Error will be added to the server log output.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:101]: This is not correct.

[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:101]: I used grammarly for this:
In the case of an incomplete response sent, an Error will be added to the server log output.

File: README.md [hunk 14]
```diff
@@ -1 +1,154 @@
...
+
+#### HEAD
+
+HEAD method is safe and idempotent. The response will indicate if the method is allowed and all resources are available.
+
+#### POST
+
+POST method is not safe and not idempotent. The JAS server will only handle POST requests that are submitted with a form, either multipart/form-data or a default form type.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:109]: I wrote it twice on purpose. I wanted to be very clear.

File: README.md [hunk 15]
```diff
@@ -1 +1,154 @@
...
+HEAD method is safe and idempotent. The response will indicate if the method is allowed and all resources are available.
+
+#### POST
+
+POST method is not safe and not idempotent. The JAS server will only handle POST requests that are submitted with a form, either multipart/form-data or a default form type. 
+In the case of a POST request for a file, the JAS Server is capable of creating the resource without a CGI script. The resource will be created in the location given in the `action` option of the html form. Otherwise, after verification of the validitiy of the request, the POST handling will be passed to the CGI Handler.
+
+#### DELETE
+
+DELETE method is not safe, but is idempotent. The JAS server will verify the request and delete the resource, similar to a command line use of `rm` if the method is allowed, etc. It is possible to pass the handling of the DELETE method to a CGI script if the user desires more fine grained control of the verification and handling of the request.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:114]: done

File: README.md [hunk 16]
```diff
@@ -1 +1,154 @@
...
+
+### CGI
+
+The Common Gateway Interface provides users of the JAS Server the ability to create srcipts to modify and create content to be returned.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:118]: I think I made the change. It is not clear what exactly you want here.

File: README.md [hunk 17]
```diff
@@ -1 +1,154 @@
...
+
+#### Limitations
+
+Due to system limitations for the project data sent to and responses returned from CGI scripts are limited to 60KB.
+In cases where a larger file is sent by POST to a cgi script the server will make a temporary file. The path to this fill can be parsed as a key=value pair from the QUERY_STRING in the the environment passed to the script. The key will be `temp_filepath` The temporary file will be deleted once the response is sent to the client.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:123]: It isn't an extra The. Just a mising period.

File: README.md [hunk 18]
```diff
@@ -1 +1,154 @@
...
+
+## How To Use
+
+The project can be compilied with `make` and ran `./webserv config_file_here`.
+If no config file is given the default designated in the `Webserv.hpp` will be used.
+- The config file can be custom and need not be one of the provided examples as long as if follows the guidelines for what is required.
+
+## Robusticity
+
+Siege was used to stress test the server. We acheived 99.8% availability.
```
[the remark this answers is not in the document]
Comment (jboucher154) [reply README.md:149]: done

**Review by sakarisalmi — APPROVED** [review]
I love it!

**Pull request merged**